Fixes pasting to existing columns, new columns and new data frame that has double quotes

Patowhiz opened the pull request:
Fixes #7095 Fixes partly #6579  Fixes #5991

rdstern reviewed (commented):
@Patowhiz as you say the file can't yet be imported.
But the earlier problem seems to have gone now - unless the file I am using now doesn't have the same quotes problem?  It now seems to import and be shown fine in the preview in the master version?

Patowhiz commented:
@rdstern yes, it can now be shown in the preview. I have fixed it in this PR. The preview would also not have worked without this fix.
Now we just need to wait for @lloyddewit response about the RScript bug. He could decide whether to merge and wait for that bug be fixed, or this PR to wait until that bug is fixed. 
Thanks.

lloyddewit commented:
@Patowhiz Please see my comment in issue #7095 

![image](https://user-images.githubusercontent.com/57253949/153861431-7cdc555e-0598-43e8-a425-074264e31697.png)

what do you think?

rdstern reviewed (commented):
I confirm that it is working ok as far as it can go.  I can't quite follow the state of the discussion on RScript - above my pay grade I think!
I still like the intention of having a Copy Special and Paste Special, but should they be disabled, or made invissible?

rdstern commented:
@Patowhiz and @lloyddewit this seems stalled.  As mentioned a month ago, I wonder what needs to be done, and by whom?

Patowhiz commented:
@rdstern all pasting done. As mentioned in the issue, I'm waiting for the RScript library bug to be fixed by @lloyddewit .
Copying and copy special still needs your input before implementation.

rdstern commented:
@Patowhiz there is a comment from @lloyddewit above that seems to imply the bug is in R-Instat?

Patowhiz commented:
@rdstern we discussed about it in the issue and @lloyddewit agreed it is in the RScript library.

lloyddewit commented:
> @rdstern we discussed about it in the issue and @lloyddewit agreed it is in the RScript library.

Correct, my comments are in issue #7095

lloyddewit commented:
@Patowhiz the RScript issue should be fixed in PR #6386.
In this PR, I don't think you can pull the changes from PR #6386 (or upgrade to RScript 1.0.5 using the NuGet package manager) without triggering merge conflicts. You may need to make a new PR to test RScript 1.0.5 with your changes.

lloyddewit commented:
> @Patowhiz the RScript issue should be fixed in PR #6386. In this PR, I don't think you can pull the changes from PR #6386 (or upgrade to RScript 1.0.5 using the NuGet package manager) without triggering merge conflicts. You may need to make a new PR to test RScript 1.0.5 with your changes.

@Patowhiz please could you look at this when you get time?
thanks

rdstern commented:
@Patowhiz is this finished to test and hopefully merge now?

rdstern commented:
@Patowhiz repeating question from above.  Is this now fixed and ready to be merged?

rdstern commented:
@Patowhiz is this ok now?

rdstern commented:
@Patowhiz any answer?  Is this ok to merge?

Patowhiz commented:
@rdstern copying of columns is still pending. This purely targets the data viewer window.

rdstern reviewed (changes requested):
@Patowhiz I thought - as we are "on a roll" for the forthcoming version, that I should look at your paste quickly.
a) Could you enable <ctrl> V  - and Edit > Paste, as you have done for the script window?  And maybe <ctrl> C and copy from the Edit menu - though they may be working already.
b) And disable the one paste that works currently - from the right-click menu.  The new version doesn't have the Paste option - and I am happy without it there.

Then I suggest it may be sufficiently useful to merge, even if there remain some options for a future issue?

rdstern commented:
@Patowhiz looking forward.  I note your message above.  There is quite a nice sequence that sort of finishes with your message:

"@rdstern I'll prioritize this and have it ready"

I write today, because it is the one-year anniversary of that message!  It's one of the items I'd like to check with you of course.

Patowhiz commented:
@N-thony yes. I'm now working on this.

Patowhiz commented:
@africanmathsinitiative/developers this is ready for review.

Patowhiz commented:
@rdstern I'm not sure on how strict we should be when it comes to pasting new columns to existing data frame. @dannyparsons leaned towards bein strict. The R function that is used here was written by him and it enforces the strictness he suggested. I'm open to amending it if you suggest so.

rdstern commented:
@Patowhiz I am excited and checking this now.  We really, really need pasting before telling users how wonderful the R-Instat spreadsheet is now.  But we also need to say what a bad idea pasting usually is, so I am very happy to remian strict.  And also R-Instat is designed so things are easy.  But I don't want pasting to be quite so easy, which is why it is not on the new right-click menu.

rdstern reviewed (commented):
@Patowhiz I think this is great.  I am not sure I am yet checking everything, but here is what I found and hence also would like.  Then I would like to merge as soon as @lloyddewit is happy.

I have just used the simple survey data, so no limits of long columns.  Questions on those can come later.
a) Ctrl+ C seems to work and so does the Edit > Copy menu item.  But ctrl + V doesn't yet work, nor does Edit > Paste.  That's all I think I need for now!  
b) Just reporting: The right-click > Paste does work, but that option is no longer on the new right-click menu.  And I am happy to not have it.  (That's part of the message - it is really easy to paste (e.g. ctrl +V , but it isn't on the toolbar!  We would like people to think before pasting.  There is usually a better way to proceed.  )
c) I tried copying some observations from yield and then pasted them lower down - warned, which is good, and then did it.  Happy with that.
d) I made a new column and pasted into it - also worked.  Good
e) I tried pasting into a factor column and it gave an error.  Very happy that it didn't work!  
f) So, see below, I made it a character variable and pasted in.  Warned, and then pasted and so messed up my data.  Very, very, happy with that!

<img width="353" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/1a40d996-c180-4593-bfe1-08387549d2e7">

So, almost ready to go.  Great!

Patowhiz commented:
@rdstern kindly test pasting new columns too. It's under the Edit > Paste New Columns. The dialog replaces pasting new data frame.

rdstern commented:
@Patowhiz this looks very impressive, but there is a lot more to report.

a) First I want to show a problem I had, parhaps unrelated, but I wanted to record it, if not.

<img width="395" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/12840627-6a01-4661-aca7-c932354374a6">

I am using the usual survey and tried a few things.  Then I got a subset, through the filter, for the first 3 villages and the result is odd.

<img width="395" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/a723047f-4b73-4ce0-b569-82dd36eeaaf3">

I checked quickly doing the same thing in Version 0.7.6 and the latest version 0.7.16, and all is fine there.  So I wanted to report that .

b) Let me also get my understanding on what we are aiming for here.  The first is to copy columns between data frames in R-Instat.  The second is to copy from (say) Excel and paste into columns or a new data frame in R-Instat.  And this option is only making new columns.  So it is a column only copier and paster.  Correct?

c) I first tried to get a copy of the whole survey data frame.  And it almost works:

<img width="451" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/37a8542a-4e44-40f4-9893-7db05bf83871">

1) The first row is header is working oddly.  It is ticked when you start, but the result is as if it is not ticked.  If you untick, and then tick again, you get the picture above, which is what I wanted, i.e. with the V1, V2, etc.  But when I pressed Ok, it ignored that, and so used the first row, and just gave me 35 rows in the new data frame.
2) In the bottom right it says Number of rows: 18.  I don't want to know the number of rows in the preview.  I want the number that will be imported. 

I think this option - As Data Frame - should soon be fine.  Great!

d) When I try As Columns I have never succeeded in getting anything except "No Data Loaded", so OK is never enabled, and I have nothing to test.

I also question what you are planning when the columns you are adding are too short, or too long, compared to the data frame.  If they are too short, you can either refuse, or fill the rest with missing values.  If too long then you can either refuse, or just stop when you get to the end of the data frame?  What is the plan?

Patowhiz commented:
@rdstern thanks for testing comprehensively. I have fixed the bugs mentioned above.

@rdstern let me clarify some points.

1. Currently we only support a maximum of 1000 rows at most. Reason being we want reproducibility. That means we have to convert the clipboard data to text then channel it to R environment as part of an R script. This makes affects performance when it comes to big datasets. If reproducibility was not a requirement, we could have supported bigger data sets.
2. @dannyparsons leaned towards being strict when pasting new columns to existing data frames. Notice I'm just using an R function that he wrote. The length being equal to the data frame. That's what I have now. If you strongly suggest that we allow flexibility in this, I can quickly implement it. 
3. Copying from any grid currently just copies the visible data (this is implemented by the `reogrid`). The is no implementation of copying the underlying data (from R). This is for performance reasons (this doesn't mean we can't implement it, if we wanted to).
4. Note, if a user tries to paste new columns that have the same name as existing, the new column names just replace the existing ones. This implementation is consistent with what @dannyparsons implemented (I'm just using the same R Function). I can change it if you want me too

On a separate note, 

1. As a user I find it very odd that when I highlight a section of columns and do a right click, I cannot find the Copy option. probably because I'm used to getting copy options in right clicks.
2. I'm also not sure if Paste New Columns is the best label for this dialog. But we can change it later. I also wonder whether we should change the 2 top radio buttons to; `As New Data Frame`, `As New Columns`. 
3. @lloyddewit I noticed that `clsRScript.GetLstTokens()` takes a lot of time to compute when I try pasting fairly many columns. In understand this is 1 scenario where a script has a dataset that has multiple lines. Is there a way for the library to quickly get comment tokens only from the script. This will improve the overall performance significantly for users that switch off showing of commands. If this can be implemented then I can make changes to `clsOutputElement` to have a function that returns comments only. I have added a comment about this in `clsOutputElement.FormattedRScript` . @rdstern expect some performance issues when pasting data sets with many columns until we fix the tokens issue with @lloyddewit.

rdstern reviewed (commented):
@Patowhiz this is really great.  It seems to be working and I love it!

I am assuming I sometimes want to paste from an R-Instat sheet or from (usually) Excel.  Here are 20 rows from Excel to go into a new sheet in R-Instat that just had 20 rows:

<img width="439" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/204b2891-fd43-424c-961d-67b8a62894c5">

a) I first got it wrong, because, in Excel I went down to line 20 - so I only had 19 rows of data.  And the paste therefore failed.
1. Could you re-instate the line in the dialog, that says how many lines there are.  And the lines to be imported, not the lines in the preview.  That would be nice and clear.
2. When imported it looked nice again - 3 figures.  I am not sure whether I like your dialog showing all the figures or not.  Maybe leave it as it is?

b) When I copied over from R-Instat - rows 1 to 20, it messed me up as you said it would, by overwriting V1 and V2 that it had used for me earlier.  That's because I can't copy the names currently in R-Instat.  I am sure we discussed that point before.  Could we copy the names from R-Instat, as we do from Excel, or is that a limitation? 
If it is a limitation, then could we at least have an up-down to start the V? from where we want?
c) I made some suggestions for minor changes in the names of the dialog and at the top.

Oooh this is really nearly ready.  Exciting!

rdstern commented:
@Patowhiz the help number for this dialog is 160.

I also note that simply adding columns side-by-side, from one data frame to another is really easy in R! But I am happy it isn't so easy in R-Instat.  Except a bit, for small problems, with Copy-Paste!  I am thinking ahead to the documentation!  

It is the bind_cols command?  I use bind_rows a lot - that's in the Data Reshape > Append dialog.  (And that is a common Copy-Paste in Excel that we don't have in R-Instat - but stack - so bind_rows is pretty easy to use.  When it doesn't work it is also for good reasons.

We deliberately don't have a bind_cols dialog, because it is risky and so we strongly prefer the Prepare >  Data Frame > Add (Merge) Columns.  That needs key columns and so is harder.  But not very hard and much less "accident-prone.

Here is a data frame where I have 2 copies of the survey, side-by-side.  

<img width="541" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/fc674cf0-1ca4-404d-b66f-53b7d1fde6bf">

It also shows what a mess you may get with the names, when they are the same!

rdstern reviewed (approved):
@Patowhiz great.  @lloyddewit over to you.

Patowhiz commented:
@rdstern I have now fixed some of the items mentioned above.

Solving the decimals issues at the preview level is not so direct. the decimals exist in the clipboard (try pasting to a notepad to see this). At the preview level, we don't have R-Instat metadata applied. I suggest leaving it that way for now as it shows the exact state of data in memory.

rdstern reviewed (approved):
@Patowhiz not sure what has changed since I last approved.  @lloyddewit I am approving again.

I did notice that the error message is a bit long and daunting.  I had commands off - which I expect from copy/pasters!  This is what I got, from just 36 rows wrongly pasted.  Could we just give the top 2 lines here, and not the pretty overwhelming wrong R command?:

<img width="503" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/ed9d0c88-a2cf-4cf0-bfb8-8c0a834789ac">

Patowhiz commented:
@rdstern currently, script error handling is generally designed to show the entire script. There is no way to disabled this at the dialog level. Not unless I implement your suggestion in a a general way. Are you happy I do this in a separate PR?

rdstern commented:
@Patowhiz very happy that it is separate and also not urgent.  So my approval still stands

lloyddewit commented on instat/dlgPasteNewColumns.vb:
```diff
@@ -1,5 +1,226 @@
+    Private bFirstLoad As Boolean = True
+    Private bReset As Boolean = True
+    'Private clsImportColsToNewDFRFunction, clsImportNewDataListRFunction As New RFunction
+    Private clsReadDataDataRFunction As New RFunction
```
rename?

Patowhiz replied:
Done.

lloyddewit commented on instat/dlgPasteNewColumns.vb:
```diff
@@ -1,5 +1,226 @@
+                Return False
+            End If
+
+            'try to show preview the data only
```
confusing comment? rephrase?

Patowhiz replied:
Done

lloyddewit commented on instat/dlgPasteNewColumns.vb:
```diff
@@ -1,5 +1,226 @@
+    End Function
+
+    Private Sub TestOkEnabled(Optional bValidateCopiedData As Boolean = True)
+        Dim enableOK As Boolean = False
+        If rdoDataFrame.Checked Then
+            enableOK = ucrSaveNewDFName.IsComplete()
+        ElseIf rdoColumns.Checked Then
+            enableOK = Not String.IsNullOrEmpty(ucrDFSelected.strCurrDataFrame)
+        End If
+        If bValidateCopiedData Then
+            enableOK = ValidateAndPreviewCopiedData()
+        End If
```
```suggestion
        Dim enableOK As Boolean = False
        If bValidateCopiedData Then
            enableOK = ValidateAndPreviewCopiedData()
        ElseIf rdoDataFrame.Checked Then
            enableOK = ucrSaveNewDFName.IsComplete()
        ElseIf rdoColumns.Checked Then
            enableOK = Not String.IsNullOrEmpty(ucrDFSelected.strCurrDataFrame)
        End If
```

Patowhiz replied:
@lloyddewit I still want to validate and preview copied data in both options if `bValidateCopiedData = True`. That's why it's the last check.

lloyddewit replied:
@Patowhiz As far as I can see, both the red and green code have the same functionality, but the red code may unnecessarily set `enableOK` twice and needs an extra line of code.
Do you have a test case where the green code fails?

The green code was just a suggestion, so if you prefer to ignore, then that's also fine from my side.

lloyddewit reviewed (changes requested):
@Patowhiz Thanks, looks good. Just some very small comments

rdstern reviewed (changes requested):
@Patowhiz I think it is probably working fine.  But <`ctrl> V` doesn't yet work and neither does `Paste` from the Edit menu.  Please could they be added?  You wer happy to add the short-cut keys into the script and it would be great to have them here too.

rdstern commented:
@Patowhiz I note that `<ctrl> C` is activated.  Whern pasting into this window what about at least activating `<ctrl> V`?

Then, with your right, I agree your Paste is looking a bit lonely.  So, if Paste remains there, then I don't have a problem adding Copy.

However, I would prefer neither!  This (of course) makes life a little harder, because some (spreadsheet) users become maniac  copy pasters.  And they often take it to extremes by doing copy-paste repeatedly.  

I would like to explain why copy-paste should never be done repeatedly (the computer simply laughs at you).  And there is usually a better way of doing it, even once!  

For example, we have included append, which is bind rows, but we don't have bind_cols which is the same as pasting extra columns - but we make that process very easy with an Add(merge).

However, any principles I have are dwarfed by my wish to complete the important addition of pasting as soon as possible.  So I am happy (at least for now) to accept Copy and Paste on the right-click menu.  I hope we can do the conventional `<ctrl> V` as well?

Patowhiz commented:
@rdstern are you now suggesting, we add Copy and Paste in the right click menu of the data viewer. Then disable the copy on the Edit menu for now?

rdstern commented:
1. And you keep ignoring my request for the short-cut of  `<ctrl>V` to facilitate pasting in the data viewer.
2. I am (slightly reluctanty) accepting Copy and Paste (hopefully with their short-cuts.) in the right-click menu of the data viewer.
3. Why do we need to disable Copy in the edit menu?

rdstern commented:
@Patowhiz I tried testing, because I saw you made some changes yesterday, but I am not sure all your final edits have been done?  For example, Paste on the Edit menu is still disabled.
Tell me when it is ready for a final test.  It would be great to include it, in the version next week.

Patowhiz commented:
@rdstern this is now ready for testing.

rdstern reviewed (dismissed):
Seems to work fine.  I used Edit > Copy and Edit > Paste.  I added some code and also made a new column and added there too.

I am approving.

a) As a separate issue could there be the option to copy from the output window?  There is no right-click there.  There is a copy button, but only if you have marked a section, so not consistent with the copy elsewhere.  Perhaps additionall there could be a right-click and ordinary ctrl C.  

b) I then tried marking a set of numbers in the log file.  They are all there!  It then has Copy (ctrl C)  etc, but they say they are disabled.  You can't paste there, but ctrl C works fine, so I am not sure why it can't be enabled.
  
@ChrisMarsh82 I assume this is you to check and merge?

Patowhiz commented:
@rdstern thanks for testing, we can have the Copy and paste of other windows done in a separate pull request.

rdstern commented:
@Patowhiz I agree that other windows - particularly the output window - is a new issue.  I have approved this one.

rdstern commented:
@ChrisMarsh82 is the checking and merging of this one for you now?  It is a 16 month old pull request, so it would be great to get it merged?

ChrisMarsh82 reviewed (dismissed):
Code looks good

Patowhiz commented:
@lloyddewit I have answered and resolved your comments. Kindly approve so that this can be merged to avoid conflicts.

Patowhiz commented:
@rdstern could you approve this then I can merge.

rdstern reviewed (approved):
@Patowhiz I am approving and great to have this done.
I thought you were going to get rid of the last entry here - the Script item, see below - as no longer needed?  This was at the same time as other small changes were being made to this menu?

<img width="391" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/ad9ddc03-8d84-4dd2-ac2a-6f905cce35e9">

Patowhiz commented:
@rdstern thanks I'll quickly remove this in another PR.
